File: assignment1/GLA.cs
// Author: Roosevelt Desouza
// Rollno: 2436
// Title: Goa Legislative Assembly OOP concepts
// Start Date: 15 July 2024
// Modified Date: 15 July 2024
// Description: To go through the Goa Legislative Assembly website and find OOP concepts and implement them in C#

using System;

class MLA
{
    // Private members for encapsulation
    private string name;
    private string constituency;
    private string party;

    // Constructor
    public MLA(string name, string constituency, string party)
    {
        this.name = name;
        this.constituency = constituency;
        this.party = party;
    }

    // Copy Constructor
    public MLA(MLA other)
    {
        this.name = other.name;
        this.constituency = other.constituency;
        this.party = other.party;
    }

    // Public methods to get the private data
    public string GetName() { return name; }
    public string GetConstituency() { return constituency; }
    public string GetParty() { return party; }

    // Public methods to set the private data
    public void SetName(string name) { this.name = name; }
    public void SetConstituency(string constituency) { this.constituency = constituency; }
    public void SetParty(string party) { this.party = party; }

    // Overriding the ToString method
    public override string ToString()
    {
        return $"MLA\nName: {name}\nConstituency: {constituency}\nParty: {party}\n";
    }
}

// Derived class using single inheritance
class ChiefMinister : MLA
{
    // Constructor calling base class constructor
    public ChiefMinister(string name, string constituency, string party) : base(name, constituency, party) {}

    // Copy Constructor
    public ChiefMinister(ChiefMinister other) : base(other) {}

    // Overriding ToString method
    public override string ToString()
    {
        return $"Chief Minister\nName: {GetName()}\nConstituency: {GetConstituency()}\nParty: {GetParty()}\n";
    }
}

class Program
{
    static void Main(string[] args)
    {
        // Creating MLA and Chief Minister objects
        MLA mla1 = new MLA("Jit Arolkar", "Mandrem", "Maharashtrawadi Gomantak");
        MLA mla2 = new MLA("Pravin Arlekar", "Pernem", "BJP");
        ChiefMinister cm = new ChiefMinister("Pramod Sawant", "Sanquelim", "BJP");

        // Copy constructor usage
        MLA copiedMLA = new MLA(mla1);

        int ch1;
        do
        {
            Console.WriteLine("Select an option:\n1: The Legislature \n2: The Government \n3: Exit");
            ch1 = int.Parse(Console.ReadLine());

            switch (ch1)
            {
                case 1:
                    int ch2;
                    do
                    {
                        Console.WriteLine("Select an option:\n1: M.L.A \n2: Exit");
                        ch2 = int.Parse(Console.ReadLine());
                        switch (ch2)
                        {
                            case 1:
                                Console.WriteLine(mla1);
                                Console.WriteLine(mla2);
                                Console.WriteLine("Copied MLA Details:");
                                Console.WriteLine(copiedMLA); // Using the copied MLA object
                                break;
                            case 2:
                                Console.WriteLine("********************Exiting********************");
                                break;
                            default:
                                Console.WriteLine("Please select a valid option.");
                                break;
                        }
                    } while (ch2 != 2);
                    break;

                case 2:
                    int ch3;
                    do
                    {
                        Console.WriteLine("Select an option:\n1: Chief Minister \n2: Exit");
                        ch3 = int.Parse(Console.ReadLine());
                        switch (ch3)
                        {
                            case 1:
                                Console.WriteLine(cm);
                                break;
                            case 2:
                                Console.WriteLine("Exiting");
                                break;
                            default:
                                Console.WriteLine("Please select a valid option.");
                                break;
                        }
                    } while (ch3 != 2);
                    break;

                case 3:
                    Console.WriteLine("Exiting");
                    break;

                default:
                    Console.WriteLine("Please select a valid option.");
                    break;
            }
        } while (ch1 != 3);
    }
}

File: assignment1/GLA.cpp
#include <iostream>
#include <string>
using namespace std;

class MLA {
private:
    string name;
    string constituency;
    string party;

public:
    MLA(string name, string constituency, string party) {
        this->name = name;
        this->constituency = constituency;
        this->party = party;
    }

    string getName() {
        return name;
    }

    string getConstituency() {
        return constituency;
    }

    string getParty() {
        return party;
    }

    void setName(string name) {
        this->name = name;
    }

    void setParty(string party) {
        this->party = party;
    }

    void setConstituency(string constituency) {
        this->constituency = constituency;
    }

    virtual void display() {
        cout << "MLA\n" 
             << "Name: " << name << "\n"
             << "Constituency: " << constituency << "\n"
             << "Party: " << party << "\n";
    }
};

class ChiefMinister : public MLA {
public:
    ChiefMinister(string name, string constituency, string party)
        : MLA(name, constituency, party) {}

    void display() override {
        cout << "Chief Minister\n"
             << "Name: " << getName() << "\n"
             << "Constituency: " << getConstituency() << "\n"
             << "Party: " << getParty() << "\n";
    }
};

int main() {
    MLA mla1("Jit Arolkar", "Mandrem", "Maharashtrawadi Gomantak");
    MLA mla2("Pravin Arlekar", "Pernem", "BJP");
    ChiefMinister cm("Pramod Sawant", "Sanquelim", "BJP");

    int ch1, ch2, ch3;
    do {
        cout << "Select an option:\n1: The Legislature \n2: The Government \n3: Exit\n";
        cin >> ch1;

        switch (ch1) {
        case 1:
            do {
                cout << "Select an option:\n1: M.L.A \n2: Exit\n";
                cin >> ch2;
                switch (ch2) {
                case 1:
                    mla1.display();
                    mla2.display();
                    break;
                case 2:
                    cout << "********************Exiting********************\n";
                    break;
                default:
                    cout << "Please select a valid option.\n";
                }
            } while (ch2 != 2);
            break;

        case 2:
            do {
                cout << "Select an option:\n1: Chief Minister \n2: Exit\n";
                cin >> ch3;
                switch (ch3) {
                case 1:
                    cm.display();
                    break;
                case 2:
                    cout << "Exiting\n";
                    break;
                default:
                    cout << "Please select a valid option.\n";
                }
            } while (ch3 != 2);
            break;

        case 3:
            cout << "Exiting\n";
            break;

        default:
            cout << "Please select a valid option.\n";
        }
    } while (ch1 != 3);

    return 0;
}
